### PR TITLE
 QgsSvgCache fetches remote SVG files in a background task 

### PR DESCRIPTION
**Pull request by nyalldawson**
Previously QgsSvgCache would often try to fetch remote images using a network request on the main thread, by calling processEvents repeatedly until the request was complete.

This caused lots of bugs, since the main thread processEvents would proceed with all kinds of stuff assuming that the svg fetch operation was complete, leading to frequent crashes and deadlocks and making remote svg use impossible (it's likely that the SVG cache remote fetching code was written in the pre-multi-threaded rendering era).

There's no way to fix this with async svg fetching - we HAVE to remove the processEvents call, and a QEventLoop won't help either (since the method may be called on the main thread). Accordingly the only solution is to fetch the requested svg in the background, and return a temporary "downloading" svg for use in the meantime. We use a QgsNetworkContentFetcherTask to do this, so it's nicely integrated with task manager.

A request task is fired up when a remote svg is requested for the first time, with the temporary 
downloading svg returned for use by the caller asynchronously. QgsSvgCache then emits the remoteSvgFetched signal when a previously requested remote SVG has been successfully fetched, triggering a map canvas redraw with the correct SVG graphic.

Fixes #18504, likely others

Here's a screencast of the task in action:

peek 2018-04-02 10-11

### CODE REVIEW

File: src/core/symbology/qgssvgcache.cpp [hunk 1]
```diff
@@ -406,77 +409,75 @@ QByteArray QgsSvgCache::getImageData( const QString &path ) const
...
 
-    // wait until the image download finished
-    // TODO: connect to the reply->finished() signal
-    while ( !reply->isFinished() )
+    QNetworkReply *reply = task->reply();
+    if ( !reply )
     {
-      QCoreApplication::processEvents( QEventLoop::ExcludeUserInputEvents, 500 );
+      // canceled
+      QMetaObject::invokeMethod( const_cast< QgsSvgCache * >( this ), "onRemoteSvgFetched", Qt::QueuedConnection, Q_ARG( QString, path ), Q_ARG( bool, false ) );
```
Comment (m-kuhn) [inline src/core/symbology/qgssvgcache.cpp]: Should this method be un-consted instead?

File: src/core/symbology/qgssvgcache.cpp [hunk 2]
```diff
@@ -406,77 +409,75 @@ QByteArray QgsSvgCache::getImageData( const QString &path ) const
...
-  {
-    QgsMessageLog::logMessage( tr( "Unexpected MIME type %1 received for %2" ).arg( contentType, path ), tr( "SVG" ) );
-    reply->deleteLater();
-    return mMissingSvg;
-  }
+    // we accept both real SVG mime types AND plain text types - because some sites
+    // (notably github) serve up svgs as raw text
+    QString contentType = reply->header( QNetworkRequest::ContentTypeHeader ).toString();
+    if ( !contentType.startsWith( QLatin1String( "image/svg+xml" ), Qt::CaseInsensitive )
+         && !contentType.startsWith( QLatin1String( "text/plain" ), Qt::CaseInsensitive ) )
```
Comment (m-kuhn) [inline src/core/symbology/qgssvgcache.cpp]: What are we protecting against with the mime-type check?

**Review by m-kuhn — APPROVED** [review]
Looks good to me

File: src/core/symbology/qgssvgcache.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/symbology/qgssvgcache.cpp]: I assume it was originally put there to protect against trying to load a png or other binary into the svg renderer

File: src/core/symbology/qgssvgcache.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/symbology/qgssvgcache.cpp]: It's called from many const methods though, and i think they are validly const. It's a fine line with caches...

**Pull request merged**